games.php

```diff
@@ -4,18 +4,25 @@
 
 include('./config.php');
 
-function order_servers($a, $b)
+function order_servers($servers)
 {
-    // Find the offset of the first alphabetic character in the server name
-    preg_match('~[a-z]~i', $a['name'], $a_start, PREG_OFFSET_CAPTURE);
-    preg_match('~[a-z]~i', $b['name'], $b_start, PREG_OFFSET_CAPTURE);
-
-    // Compare the alphabetic portion only, ignoring special characters at the start
-    // Servers without any alphabetic characters are pushed to the end
-    $a_comp = $a_start ? substr($a['name'], $a_start[0][1]) : "z" . $a['name'];
-    $b_comp = $b_start ? substr($b['name'], $b_start[0][1]) : "z" . $b['name'];
-
-    return strcasecmp($a_comp, $b_comp);
+	// Create an array of server names, with any non-alphanumeric characters replaced with z
+	$names = array();
+	foreach($servers as $key => $server){
+		$names[$key] = strtolower(preg_replace("/[^\p{L}|\p{N}]/u", 'z', $server['name']));
```

Comment:
According to http://www.regular-expressions.info/unicode.html `\p{L}` will only match a single code point, ignoring diacritics or other combining marks.  I'm not familiar with how Collator works, so this may be intentional?  According to the same link you can use `\p{L}\p{M}*+` to include combining marks.

Reply:
This line isn't to do with the Collator, it's just using a regular expression to replace anything that doesn't match /[^\p{L}|\p{N}]/u with a z.

/     /u - Contains unicode regular expression
[^   ] - Matches anything not contained within
\p{L} Matches any kind of letter from any language
\p{N} Matches any kind of numeric character in any script

Reply:
Yup, but `\p{L}` extracts only the letter mark, and so something like `àb` will be cleaned to `azb` according to the link above. I was not sure whether this was working around a limitation of `Collator->asort` or was just an oversight.

Reply:
I see what you are saying, although àb wont be cleaned at all because both characters are alphanumeric.

Reply:
I've tested replacing multibyte puntuation with 'z' and it seems to work fine. I'm not sure how long it lasts, but I have a test bed here:

https://www.tehplayground.com/0366VSS1q2UySbwR

The character I'm test with is '，' which encodes to EF BC 8C.

Reply:
It depends on how it is encoded. `à` was a bad example because it has two different encodings.  The single-codepoint version is a special case that is alphanumeric, but the two-codepoint version only considers the first while (correctly) not the combining mark:

```php
<?php
function filter($x)
{
    return strtolower(preg_replace("/[^\p{L}|\p{N}]/u", 'z', $x));
}

$a = "\u{00E0}b";
print("Single codepoint encoding is fine: " . $a . " -> " . filter($a) . "\n");

$b = "a\u{0300}b";
print("Letter plus diacritic isn't: " . $b . " -> " . filter($b) . "\n");
?>
```
```
paul@ubuntu-vm:~$ php test.php
Single codepoint encoding is fine: àb -> àb
Letter plus diacritic isn't: àb -> azb

```

Reply:
The case i'm worried about is comparing e.g. "a great server" vs "à bad server" .  IMO the sort order should place "bad" first, but that won't happen if `à` is normalized to `az`.

Reply:
It isn't. I dont think your strings are being produced correctly, if you just use À rather than trying to add the accent manually, you will see that no 'z's are added.

Reply:
Ok I've made it normalize strings before the regular expression to prevent this issue.

Reply:
Ok this looks more reasonable.  If we run into problems with languages that do not have combined forms in the future then we can revisit this then.  I have run out of time to test this further tonight, but will try and get back to it in the next couple of days.